### PR TITLE
(flw 2407) scan wallet qr

### PR DESCRIPTION
**Pull request by Viscount88**

### CODE REVIEW

File: fearless/Modules/WalletMainContainer/WalletMainContainerPresenter.swift [hunk 1]
```diff
@@ -152,3 +159,42 @@ extension WalletMainContainerPresenter: SelectNetworkDelegate {
...
+            delegate: self
+        )
+    }
+
+    func didFinishWith(address _: String) {}
+}
+
+extension WalletMainContainerPresenter: SendPrepareUseCaseDelegate {
+    func didReceive(possibleChains: [ChainModel]) {
+        if possibleChains.isEmpty {
```
Comment (DRadmir) [inline fearless/Modules/WalletMainContainer/WalletMainContainerPresenter.swift:176]:
```suggestion
        guard possibleChains.isNotEmpty else { return }
```
Thus there will be no investment else if
(optional)

File: fearlessTests/Modules/ScanQR/ScanQRTests.swift [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import XCTest
```
Comment (DRadmir) [inline fearlessTests/Modules/ScanQR/ScanQRTests.swift:1]: remove empty test

File: fearless/Modules/NewWallet/SendFlow/ChooseRecipient/ChooseRecipientPresenter.swift [hunk 3]
```diff
@@ -31,6 +32,10 @@ final class ChooseRecipientPresenter {
         self.wallet = wallet
         self.qrParser = qrParser
         self.localizationManager = localizationManager
+
+        if let address = address {
+            view?.didReceive(address: address)
```
[the remark this answers is not in the document]
Comment (Viscount88) [reply fearless/Modules/NewWallet/SendFlow/ChooseRecipient/ChooseRecipientPresenter.swift]: I removed this module, it was temporary solution

File: fearless/Modules/ScanQR/ScanQRAssembly.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    static func configureModule(moduleOutput: ScanQRModuleOutput) -> ScanQRModuleCreationResult? {
+        let localizationManager = LocalizationManager.shared
+
+        let qrDecoder = NewQRCoderFactory().createDecoder()
+        let qrScanMatcher = QRScanMatcher(decoder: qrDecoder)
+
+        let qrScanService = QRCaptureServiceFactory().createService(
+            with: qrScanMatcher,
+            delegate: nil,
+            delegateQueue: nil
```
[the remark this answers is not in the document]
Comment (Viscount88) [reply fearless/Modules/ScanQR/ScanQRAssembly.swift:14]: It's okay

File: fearless/Modules/ScanQR/ScanQRInteractor.swift [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+        qrExtractionService.extract(
+            from: image,
+            using: matcher,
+            dispatchCompletionIn: .main
+        ) { [weak self] result in
+            switch result {
+            case .success:
+                if let addressInfo = matcher.addressInfo {
+                    self?.output?.handleMatched(addressInfo: addressInfo)
```
[the remark this answers is not in the document]
Comment (Viscount88) [reply fearless/Modules/ScanQR/ScanQRInteractor.swift:41]: We handle it like error

File: fearless/Modules/ScanQR/ScanQRRouter.swift [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        case .denied:
+            delegate.didFail(in: self, with: ImageGalleryError.accessDeniedPreviously)
+        default:
+            presentGallery(from: view, delegate: delegate, pickerDelegate: pickerDelegate)
+        }
+    }
+
+    private func presentGallery(
+        from view: ControllerBackedProtocol?,
+        delegate _: ImageGalleryDelegate,
```
[the remark this answers is not in the document]
Comment (Viscount88) [reply fearless/Modules/ScanQR/ScanQRRouter.swift:33]: Are you sure?)

File: fearless/Modules/WalletMainContainer/UseCases/SendPrepareUseCase.swift [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import RobinHood
+
+protocol SendPrepareUseCaseDelegate: NSObject {
+    func didReceive(possibleChains: [ChainModel])
+    func didReceive(chainAsset: ChainAsset, address: String)
+}
+
+final class SendPrepareUseCase {
+    private let chainModelRepository: AnyDataProviderRepository<ChainModel>
+    private let operationQueue = OperationQueue()
```
[the remark this answers is not in the document]
Comment (Viscount88) [reply fearless/Modules/WalletMainContainer/UseCases/SendPrepareUseCase.swift]: This use case removed

File: fearless/Modules/WalletMainContainer/WalletMainContainerPresenter.swift [hunk 8]
```diff
@@ -1,7 +1,8 @@
 import Foundation
 import SoraFoundation
+import FearlessUtils
 
-final class WalletMainContainerPresenter {
+final class WalletMainContainerPresenter: NSObject {
```
[the remark this answers is not in the document]
Comment (Viscount88) [reply fearless/Modules/WalletMainContainer/WalletMainContainerPresenter.swift:5]: use case was

File: fearless/Modules/WalletMainContainer/WalletMainContainerPresenter.swift [hunk 9]
```diff
@@ -159,3 +161,21 @@ extension WalletMainContainerPresenter: SelectNetworkDelegate {
         interactor.saveChainIdForFilter(chain?.chainId)
     }
 }
+
+extension WalletMainContainerPresenter: ScanQRModuleOutput {
```
Comment (DRadmir) [inline fearless/Modules/WalletMainContainer/WalletMainContainerPresenter.swift:165]: methods in this protools do the same. Maybe remove didFinishWith(addressInfo: AddressQRInfo)

File: fearless/en.lproj/Localizable.strings [hunk 10]
```diff
@@ -895,3 +895,4 @@ Remember to make a backup of your key and keep it in a safe and private place (e
 "all.done.alert.hash.stub" = "Hash";
 "all.done.alert.result.stub" = "Result";
 "all.done.alert.success.stub" = "Success";
+"contacts.scan" = "Scan QR code";
```
Comment (DRadmir) [inline fearless/en.lproj/Localizable.strings:898]: seems to have been changed to short? "QR-code"?

**Review by DRadmir — CHANGES_REQUESTED** [review]

File: fearless/Common/QRCoding/QRCoderFactory.swift [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+        guard let info = info as? AddressQRInfo else {
+            throw QRDecoderError.wrongDecoder
+        }
+
+        return info
+    }
+}
+
+final class NewAddressQRDecoder: QRDecodable {
```
[the remark this answers is not in the document]
Comment (Viscount88) [reply fearless/Common/QRCoding/QRCoderFactory.swift:52]: It's temporary, when we'll sure about it's reliability we remove this class to utils

File: fearless/en.lproj/Localizable.strings (hunk 10 above)
Comment (Viscount88) [reply fearless/en.lproj/Localizable.strings:898]: I just use design description

**Review by DRadmir — APPROVED** [review]

**Pull request merged**